pkg/dao/rpm.go

```diff
@@ -170,3 +199,220 @@ func (r rpmDaoImpl) Search(orgID string, request api.SearchRpmRequest, limit int
+		Where("checksum in (?)", checksums).
+		Model(&models.Rpm{}).
+		Pluck("uuid", &rpmUuids).Error; err != nil {
+		return rowsAffected, fmt.Errorf("failed retrieving rpms.uuid for the package checksums: %w", err)
+	}
+
+	// Delete Rpm and RepositoryRpm entries we don't need
+	if err = r.deleteUnneeded(repo, rpmUuids); err != nil {
+		return rowsAffected, fmt.Errorf("failed to deleteUnneeded: %w", err)
+	}
+
+	//Add the RepositoryRpm entries we do need
+	associations := prepRepositoryRpms(repo, rpmUuids)
+	if err = r.db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "repository_uuid"}, {Name: "rpm_uuid"}},
+		DoNothing: true}).
+		Create(&associations).
+		Error; err != nil {
+		return rowsAffected, fmt.Errorf("failed to Create: %w", err)
+	}
+
+	return rowsAffected, err
+}
+
+// prepRepositoryRpms  converts a list of rpm_uuids to a list of RepositoryRpm Objects
+func prepRepositoryRpms(repo models.Repository, rpm_uuids []string) []models.RepositoryRpm {
+	repoRpms := make([]models.RepositoryRpm, len(rpm_uuids))
+	for i := 0; i < len(rpm_uuids); i++ {
+		repoRpms[i].RepositoryUUID = repo.UUID
+		repoRpms[i].RpmUUID = rpm_uuids[i]
+	}
+	return repoRpms
+}
+
+// difference returns the difference between arrays a and b   (a - b)
+func difference(a, b []string) []string {
+	mb := make(map[string]struct{}, len(b))
+	for _, x := range b {
+		mb[x] = struct{}{}
+	}
+	var diff []string
+	for _, x := range a {
+		if _, found := mb[x]; !found {
+			diff = append(diff, x)
+		}
+	}
+	return diff
+}
+
+// deleteUnneeded Removes any RepositoryRpm entries that are not in the list of rpm_uuids
+func (r rpmDaoImpl) deleteUnneeded(repo models.Repository, rpm_uuids []string) error {
+	//First get uuids that are there:
+	var (
+		existing_rpm_uuids []string
+		dangling_rpm_uuids []string
+	)
+
+	// Read existing rpm_uuid associated to repository_uuid
+	if err := r.db.Model(&models.RepositoryRpm{}).
+		Where("repository_uuid = ?", repo.UUID).
+		Pluck("rpm_uuid", &existing_rpm_uuids).
+		Error; err != nil {
+		return err
+	}
+
+	rpmsToDelete := difference(existing_rpm_uuids, rpm_uuids)
+
+	// Delete the many2many relationship for the unneeded rpms
+	if err := r.db.
+		Unscoped().
+		Where("repositories_rpms.repository_uuid = ?", repo.UUID).
+		Where("repositories_rpms.rpm_uuid in (?)", rpmsToDelete).
+		Delete(&models.RepositoryRpm{}).
+		Error; err != nil {
+		return err
+	}
+
+	// Retrieve dangling rpms.uuid
+	if err := r.db.
+		Model(&models.Rpm{}).
+		Where("repositories_rpms is NULL").
+		Where("rpms.uuid in (?)", rpmsToDelete).
+		Joins("left join repositories_rpms on rpms.uuid = repositories_rpms.rpm_uuid").
+		Pluck("rpms.uuid", &dangling_rpm_uuids).
+		Error; err != nil {
+		return err
+	}
+
+	if len(dangling_rpm_uuids) == 0 {
+		return nil
+	}
+
+	// Remove dangling rpms
+	if err := r.db.
+		Unscoped().
+		Where("rpms.uuid in (?)", dangling_rpm_uuids).
+		Delete(&models.Rpm{}).
+		Error; err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
I think it might make more sense to just leave dangling rpms within the db at introspection time and remove them either at the end of the process, or some other time.   Lets leave this for now and i filed:  https://issues.redhat.com/browse/HMSCONTENT-144